axelrod/match.py

```diff
@@ -35,8 +40,17 @@ def __init__(self, players, turns, game=None, deterministic_cache=None,
             The default is to use the correct values for turns, game and noise
             but these can be overridden if desired.
         """
-        self.result = []
+
+        if prob_end is not None:
+            if turns is not None:
+                turns = min(turns, sample_length(prob_end))
+            else:
+                turns = sample_length(prob_end)
+        elif turns is None:
+            turns = 20
```

Comment:
I've just realised I need to move this logic anyway to the `play` method (for the cache to work as we'd want), I'll see if I see a way of getting a dict to work there (although it'd need to be different to this so that it avoids unnecessary random samples). :+1: